### ISSUE TITLE
Block instead of error when mailbox full?

### ISSUE THREAD

**Issue by StefanBertels** [louthy/echo-process#23]
Is there a way to block on `tell` when posting a message to a full mailbox instead of throwing an error?

I want to use an actor for slow background processes. TPL has an ActionBlock<T> where you can set BoundedCapacity to reduce memory usage due to slow processing.

**Comment by louthy** [comment]
Sorry, missed this.  I think it's quite a dangerous function to add, so I think you'd need to implement it yourself:
```c#
   public static Unit blockingTell<A>(ProcessId pid, A msg)
   {
       while(true)
       {
           try
           {
               return tell(pid, msg);
           }
           catch(ProcessInboxFullException)
           {}
       }
   }
```

**Comment by StefanBertels** [comment]
Ok.

Scenario: actor 1 crawls the web and tells actor 2 all URLs for downloading. Actor 1 is faster and should be throttled (even when 2 is working in parallel) to avoid outofmemory.

Maybe there is a chain of intermediate actors between 1 and 2 (filtering...) to make scenario worse...

Is there a better scenario than blocking everything?

**Comment by louthy** [comment]
It's not clear from your scenario why the web-crawler should go faster than the web-downloader.  What benefit does that have?  If you have a web-crawler that can crawl the entire internet then your web-downloader will never catch-up.  If the only usage for the collected links is to subsequently download them, why not combine the crawler and downloader?  You could shard your crawler/downloader by site name for multiple parallel crawler/downloaders.

Ultimately what you're describing is having a way to deal with back-pressure which could be done in an actor system, but would be quite messy I think.

**Comment by StefanBertels** [comment]
Thanks for your answer(s), helps me a lot to understand how things should be done "right".

I really think that actors would fit better than my current OOP processing chain (even when pimped with LanguageExt). I want to split this chain into actors so that every ressouce and every transactional state is controlled by a different actor.

Yes, back-pressure is the point. My real application does not download the web, it's more complicated. 
The point is I can control speed of putting messages (e.g. URLs) into the processing chain and I have some bottleneck later. Therefore I thought about TPL and blocking inboxes.

I think actor system would be a better design. Maybe I should have some supervision actor spawning the actors for the processing chain that checks inbox size of those bottleneck actors and use that information to control the "input" actors speed. But feels like a ugly hack.

And I will get difficulties if I replay all messages from the past. This might be fast (every actors just checks, that he has already done action) but if this check is a little bit faster in a later actor I will get "inbox full" error, too. 

Comparing with TPL I'm in favor for actors, especially echo-process, because they have better design regarding state handling, setup/dispose of actors and better overall control (error handling). And in fact TPL's blocking feature could be done with blockingTell or just echo's error handling (retry) feature.

**Comment by StefanBertels** [comment]
Is there something (or could there be) in echo actor system that fits Cold Observable pattern? In fact those message I want to replay can be represented by an Enumerator. 

What I need here is some router actor that would pass hot events just 1:1 to bottlenack actor and when given some Enumerator of those events would pass them in a controlled way to the bottleneck actor.

I saw your broadcastMapMany router feature which goes into this direction. But would be better if the blocking semaphore of the bottleneck actor would be visible/usable for the router to avoid buffering. In fact passing 1:1 might be seen as special case (IEnumerable of size 1).

Or could there be something like tellMany(processId, IEnumerable<Messagetype>)?

**Comment by StefanBertels** [comment]
I will solve this by message type Seq<A> with SeqOne(message) for single messages (hot) and sending a Seq<a>(... enumerable) for things like replaying messages (cold). This just means that inbox function will fold this messages (same way whole actor folds messages with regard to state).

```csharp
public static ProcessId spawnFolding<S, T>(ProcessName name, Func<S> setup, Func<S, T, S> inbox)
    => spawn(name, setup, fold_Seq(inbox));

public static Unit tellOne<T>(ProcessId id, T oneMessage)
    => tell(id, SeqOne(oneMessage));

public static Unit tellSeq<T>(ProcessId id, T oneMessage, T secondMessage, params T[] messages)
    => tell(id, oneMessage.Cons(secondMessage.Cons(messages)));

public static Unit tellSeq<T>(ProcessId id, IEnumerable<T> messages)
    => tell(id, messages.ToSeq());

public static Func<S, Seq<T>, S> fold_Seq<S, T>(Func<S, T, S> foldOne)
    => (state, seq) => seq.Fold(state, foldOne);

[Fact]
public static void Test()
{
    ProcessConfig.initialise();
    var actor1 = spawn("actor1", Setup, fold_Seq<string, int>(Inbox));
    tell(actor1, SeqOne(42));
    tellOne(actor1, 42);
    var actor2 = spawnFolding<string, int>("actor2", Setup, Inbox);
    tellSeq(actor2, 23, 1984);
    tellSeq(actor2, Seq(1, 2, 3));
    shutdownAll();
}

public static string Setup()
{
    return "0";
}

public static string Inbox(string prev, int newValue)
{
    var newState = $"{prev}+{newValue}";
    Console.WriteLine($"{Self.Name}: ({prev}, {newValue}) => {newState}");
    return newState;
}
```

Of course large sequences might block actor which might be a problem on shutdown.

Question: Is there some CancellationToken or other solution (Self.Cancelled?) that could be checked during longer running inbox function?

**Comment by louthy** [comment]
There isn't a Self.Cancelled. In theory the process can't be cancelled
until the current message has been processed. However, because there's a
separate queue for system messages, it is possible for the Process to be
removed as a child of its parent whilst a message is being processed.

The removal of the child process by its parent also completes the publish
queues, so in theory you can wait for the OnComplete RX event.

Not at a computer right now, so you'll need to experiment:

var cancelled = false;
using( var x = observe<Unit>.Subscribe(() => cancelled = true)
{
// ... do your long task here ...
}

**Comment by StefanBertels** [comment]
Note: Thanks to #40 my last comment can be ignored (should work with observeUnsafe).
